subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_cellconn_obj.c

```diff
@@ -0,0 +1,660 @@
+{
+	int err = 0;
+	bool rai_enabled = false;
+
+	LOG_DBG("RAI value: %d", *data);
+
+	switch (*data & 7U) {
+	case 1:
+		break;
+	case 2:
+		rai_enabled = true;
+		err = lte_lc_rai_param_set("4");
+		break;
+	case 6:
+		rai_enabled = true;
+		err = lte_lc_rai_param_set("3");
+		break;
+	default:
+		LOG_ERR("Invalid RAI param");
+		return -EINVAL;
+		;
+	}
+
+	if (err) {
+		LOG_ERR("RAI set error %d", err);
+		return err;
+	}
+
+	err = lte_lc_rai_req(rai_enabled);
+	if (err) {
+		LOG_ERR("RAI req error %d", err);
+		return err;
+	}
+
+	return 0;
+}
+#endif
+
+static void psm_update(struct lte_lc_psm_cfg psm_cfg)
+{
+	if (psm_cfg.tau != tau) {
+		tau = psm_cfg.tau;
+		cache_psm_time_str(psm_cfg.tau);
+	}
+
+	if (psm_cfg.active_time != active_time) {
+		active_time = psm_cfg.active_time;
+		cache_active_time_str(psm_cfg.active_time);
+	}
+
+	if (psm_cfg.active_time > 0) {
+		active_psm_modes |= 1U;
+	} else {
+		active_psm_modes &= ~1U;
+	}
+}
+
+static void edrx_update(struct lte_lc_edrx_cfg edrx_cfg)
+{
+	uint8_t edrx;
+	uint8_t ptw;
+	uint8_t val;
+	uint8_t idx = 0;
+
+	/* Lookup table for eDRX WB-S1 mode */
+	static const uint8_t edrx_lookup_ltem[16] = {
+		0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 13, 13
+	};
+	/* Lookup table for eDRX NB-S1 mode */
+	static const uint8_t edrx_lookup_nbiot[16] = {
+		2, 2, 2, 3, 2, 5, 2, 2, 2, 9, 10, 11, 12, 13, 14, 15
+	};
+	static const uint16_t multipliers[16] = {
+		0, 1, 2, 4, 6, 8, 10, 12, 14, 16, 32, 64, 128, 256, 512, 1024
+	};
+
+	/* The eDRX value is a multiple of 10.24 seconds, except for the
+	 * special case of idx == 0 for LTE-M, where the value is 5.12 seconds.
+	 */
+	if (edrx_cfg.edrx > 10.0) {
+		val = round(edrx_cfg.edrx / 10.24);
+		for (int i = 1; i < 16; i++) {
+			if (val == multipliers[i]) {
+				idx = i;
+				break;
+			}
+		}
+	}
+
+	/* Convert eDRX and PTW value in seconds back to bits as described in
+	 * [3GPP-TS_24.008, clause 10.5.5.32]
+	 */
+	if (edrx_cfg.mode == LTE_LC_LTE_MODE_LTEM) {
+		ptw = round(edrx_cfg.ptw / 1.28) - 1;
+		edrx = edrx_lookup_ltem[idx];
+	} else {
+		ptw = round(edrx_cfg.ptw / 2.56) - 1;
+		edrx = edrx_lookup_nbiot[idx];
+	}
+
+	edrx |= (ptw << 4U);
+
+	LOG_DBG("EDRX: 0x%x", edrx);
+	if (active_time > 0) {
+		active_psm_modes = PSM_AND_EDRX_MODE;
+	} else {
+		active_psm_modes = EDRX_ONLY_MODE;
+	}
+
+	if (edrx_cfg.mode == LTE_LC_LTE_MODE_LTEM) {
+		edrx_wbs1 = edrx;
+	} else {
+		edrx_nbs1 = edrx;
+	}
+}
+
+static void lte_event_handler(const struct lte_lc_evt *const evt)
+{
+	switch (evt->type) {
+	case LTE_LC_EVT_PSM_UPDATE:
+		psm_update(evt->psm_cfg);
+		break;
+	case LTE_LC_EVT_EDRX_UPDATE:
+		LOG_DBG("LTE EDRX update: mode %s, edrx %.2f s, Paging time %.2f s",
+			(evt->edrx_cfg.mode == 7 ? "LTE" : "NBIOT"), evt->edrx_cfg.edrx,
+			evt->edrx_cfg.ptw);
+		edrx_update(evt->edrx_cfg);
+		break;
+	case LTE_LC_EVT_LTE_MODE_UPDATE:
+		LOG_DBG("LTE mode update %d", evt->lte_mode);
+		lte_mode = evt->lte_mode;
+		break;
```

Comment:
We are updating certain resources here, at least in `psm_update()` as well as `edrx_update()` without telling the engine that the data have changed.

Think we should either use `lwm2m_set_XXX()` family functions, or directly use `lwm2m_notify_observer_path()`. Otherwise if server is observing this object, it would not get the information when data have changed.

Reply:
I fixed the notification issue using lwm2m_notify_observer() when resource data changes